libvast/src/system/eraser.cpp

```diff
@@ -14,80 +14,102 @@
-  auto result = super::status(v);
+record eraser_state::status(status_verbosity) const {
+  auto result = record{};
   result["query"] = query_;
   result["interval"] = interval_;
   return result;
 }
 
-caf::behavior
-eraser(caf::stateful_actor<eraser_state>* self, caf::timespan interval,
-       std::string query, index_actor index) {
-  VAST_TRACE_SCOPE("{} {} {} {} {}", VAST_ARG(*self), VAST_ARG(interval),
-                   VAST_ARG(query), VAST_ARG(index));
-  auto& st = self->state;
-  st.init(interval, std::move(query), std::move(index));
-  return st.behavior();
+eraser_actor::behavior_type
+eraser(eraser_actor::stateful_pointer<eraser_state> self,
+       caf::timespan interval, std::string query, index_actor index) {
+  VAST_TRACE_SCOPE("eraser: {} {} {} {}", VAST_ARG(self->id()),
+                   VAST_ARG(interval), VAST_ARG(query), VAST_ARG(index));
+  // Set member variables.
+  self->state.interval_ = interval;
+  self->state.query_ = std::move(query);
+  self->state.index_ = std::move(index);
+  self->delayed_send(self, interval, atom::ping_v);
+  return {
+    [self](atom::ping) {
+      self
+        ->request(static_cast<eraser_actor>(self), self->state.interval_,
+                  atom::run_v)
+        .then(
+          [self](atom::ok) {
+            VAST_VERBOSE("{} successfully finishes run", *self);
+            self->delayed_send(static_cast<eraser_actor>(self),
+                               self->state.interval_, atom::ping_v);
+          },
+          [self](const caf::error& e) {
+            VAST_WARN("{} encountered error while erasing: {}", *self, e);
+            self->delayed_send(static_cast<eraser_actor>(self),
+                               self->state.interval_, atom::ping_v);
+          });
+    },
+    [self](atom::run) -> caf::result<atom::ok> {
+      auto const& query = self->state.query_;
+      VAST_VERBOSE("{} runs with query {}", *self, query);
+      auto expr = to<expression>(query);
+      if (!expr)
+        return caf::make_error(ec::invalid_query, fmt::format("{} failed to "
+                                                              "parse query {}",
+                                                              *self, query));
+      if (expr = normalize_and_validate(std::move(*expr)); !expr)
+        return caf::make_error(
+          ec::invalid_query,
+          fmt::format("{} failed to normalize and validate {}", *self, query));
+      auto transform
+        = std::make_shared<vast::transform>("eraser_transform", std::nullopt);
+      auto select_config = select_step_configuration{
+        .expression = self->state.query_,
+        .invert = true,
+      };
+      transform->add_step(std::make_unique<select_step>(select_config));
+      auto rp = self->make_response_promise<atom::ok>();
+      self->request(self->state.index_, caf::infinite, atom::resolve_v, *expr)
+        .then(
+          [self, transform, rp = rp](catalog_result& result) mutable {
```

Comment:
Why the `rp = rp` over just `rp` here and a few times below?

Reply:
Good question, there used to be a mysterious compiler error if I didn't, but I can't reproduce.